### DISCUSSION TITLE
为什么OnResponseBody方法中可以使用 proxywasm.ReplaceHttpRequestHeaders？

### DISCUSSION THREAD

**Discussion by mark8s** [alibaba/higress#1268] [General]
我有个疑问，我在几个Ai插件中，如 [ai-transformer](https://github.com/alibaba/higress/tree/main/plugins/wasm-go/extensions/ai-transformer) 和 [ai-security-guard](https://github.com/alibaba/higress/tree/main/plugins/wasm-go/extensions/ai-security-guard)，都看到这种写法，如下:

在onHttpResponseBody方法中使用`proxywasm.ReplaceHttpResponseHeaders(reconvertHeaders(hdsMap))`出处 ai-security-guard ：
```
func onHttpResponseBody(ctx wrapper.HttpContext, config AISecurityConfig, body []byte, log wrapper.Log) types.Action {
						hdsMap := ctx.GetContext("headers").(map[string][]string)
						delete(hdsMap, "content-length")
						hdsMap[":status"] = []string{"403"}
						**proxywasm.ReplaceHttpResponseHeaders(reconvertHeaders(hdsMap))**
						proxywasm.ReplaceHttpResponseBody(jsonData)
						proxywasm.SetProperty([]string{"risklabel"}, []byte(respResult.Array()[0].Get("Label").String()))
}
```

在onHttpResponseBody方法中`proxywasm.ReplaceHttpResponseHeaders(newHeaders)`出处 ai-transformer：
```
func onHttpResponseBody(ctx wrapper.HttpContext, config AITransformerConfig, body []byte, log wrapper.Log) types.Action {
			newHeaders, newBody, err := extraceHttpFrame(gjson.GetBytes(responseBody, "output.text").String())
			if err == nil {
				**proxywasm.ReplaceHttpResponseHeaders(newHeaders)**
				proxywasm.ReplaceHttpResponseBody(newBody)
			}
			proxywasm.ResumeHttpResponse()
}
```

但是 proxywasm.ReplaceHttpResponseHeaders() ,它难道不是只能用于 OnHttpResponseHeaders 阶段吗？ 写在这里难道没问题吗？

这是它的说明：
```
// ReplaceHttpResponseHeaders is used for replacing HTTP response headers
// with given headers. Only available during
// types.HttpContext.OnHttpResponseHeaders.
func ReplaceHttpResponseHeaders(headers [][2]string) error {
	return setMap(internal.MapTypeHttpResponseHeaders, headers)
}
```

有知道的吗？理论上这样执行并不会生效，上面也并不没有处理异常，如果处理了异常应该会报错才对。

**Comment by CH3CHO** [comment]
目测题目里的 `ReplaceHttpRequestHeaders` 是写错了，你应该是想写 `proxywasm.ReplaceHttpResponseHeaders`。

之所以这两个插件可以在 `onResponseBody` 里调用 `proxywasm.ReplaceHttpResponseHeaders`，是因为它们在 `onResponseHeaders` 里返回了 `types.HeaderStopIteration`（[参考代码](https://github.com/alibaba/higress/blob/main/plugins/wasm-go/extensions/ai-security-guard/main.go#L205)）。这是 Higress 做的一个改动，如果这样返回的话，那么可以在 `onBody` 阶段修改 headers。但要求在编译插件的时候增加一个额外的标记 `proxy_wasm_version_0_2_100`（[参考代码](https://github.com/alibaba/higress/blob/main/plugins/wasm-go/extensions/ai-security-guard/.buildrc)）。
